osu.Game/Beatmaps/WorkingBeatmap.cs

```diff
@@ -83,55 +83,73 @@ protected virtual Track GetVirtualTrack(double emptyLength = 0)
-                mod.ApplyToBeatmapConverter(converter);
+                // Check if the beatmap can be converted
+                if (Beatmap.HitObjects.Count > 0 && !converter.CanConvert())
+                    throw new BeatmapInvalidForRulesetException($"{nameof(Beatmaps.Beatmap)} can not be converted for the ruleset (ruleset: {ruleset.InstantiationInfo}, converter: {converter}).");
 
-            // Convert
-            IBeatmap converted = converter.Convert();
+                // Apply conversion mods
+                foreach (var mod in mods.OfType<IApplicableToBeatmapConverter>())
+                {
+                    cancellationSource.Token.ThrowIfCancellationRequested();
+                    mod.ApplyToBeatmapConverter(converter);
+                }
 
-            // Apply difficulty mods
-            if (mods.Any(m => m is IApplicableToDifficulty))
-            {
-                converted.BeatmapInfo = converted.BeatmapInfo.Clone();
-                converted.BeatmapInfo.BaseDifficulty = converted.BeatmapInfo.BaseDifficulty.Clone();
+                // Convert
+                IBeatmap converted = converter.Convert();
 
-                foreach (var mod in mods.OfType<IApplicableToDifficulty>())
-                    mod.ApplyToDifficulty(converted.BeatmapInfo.BaseDifficulty);
-            }
+                // Apply difficulty mods
+                if (mods.Any(m => m is IApplicableToDifficulty))
+                {
+                    converted.BeatmapInfo = converted.BeatmapInfo.Clone();
+                    converted.BeatmapInfo.BaseDifficulty = converted.BeatmapInfo.BaseDifficulty.Clone();
 
-            IBeatmapProcessor processor = rulesetInstance.CreateBeatmapProcessor(converted);
+                    foreach (var mod in mods.OfType<IApplicableToDifficulty>())
+                    {
+                        cancellationSource.Token.ThrowIfCancellationRequested();
+                        mod.ApplyToDifficulty(converted.BeatmapInfo.BaseDifficulty);
+                    }
+                }
 
-            processor?.PreProcess();
+                IBeatmapProcessor processor = rulesetInstance.CreateBeatmapProcessor(converted);
 
-            // Compute default values for hitobjects, including creating nested hitobjects in-case they're needed
-            foreach (var obj in converted.HitObjects)
-                obj.ApplyDefaults(converted.ControlPointInfo, converted.BeatmapInfo.BaseDifficulty);
+                processor?.PreProcess();
 
-            foreach (var mod in mods.OfType<IApplicableToHitObject>())
-            {
+                // Compute default values for hitobjects, including creating nested hitobjects in-case they're needed
                 foreach (var obj in converted.HitObjects)
-                    mod.ApplyToHitObject(obj);
-            }
+                {
+                    cancellationSource.Token.ThrowIfCancellationRequested();
```

Comment:
sure we don't want to re-throw this as a `TimeoutException`? Seems a bit weird for callers to handle `OperationCancelledException` when they don't pass in a token.